File: train_notify/integrations/tmux.py
from __future__ import annotations

import shlex
import shutil
import subprocess
import time
from typing import Optional


def has_tmux() -> bool:
    return shutil.which("tmux") is not None


def session_exists(session: str) -> bool:
    proc = subprocess.run(
        ["tmux", "has-session", "-t", session],
        check=False,
        capture_output=True,
        text=True,
    )
    return proc.returncode == 0


def start_detached_session(session: str, command: str, cwd: Optional[str] = None) -> None:
    shell_cmd = command
    if cwd:
        shell_cmd = f"cd {shlex.quote(cwd)} && {command}"
    proc = subprocess.run(
        ["tmux", "new-session", "-d", "-s", session, shell_cmd],
        check=False,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or "failed to start tmux session")


def send_ctrl_c(session: str) -> bool:
    proc = subprocess.run(
        ["tmux", "send-keys", "-t", session, "C-c"],
        check=False,
        capture_output=True,
        text=True,
    )
    return proc.returncode == 0


def stop_session(session: str, grace_seconds: float = 3.0) -> None:
    send_ctrl_c(session)
    start = time.monotonic()
    while time.monotonic() - start <= grace_seconds:
        if not session_exists(session):
            return
        time.sleep(0.2)
    subprocess.run(["tmux", "kill-session", "-t", session], check=False)

File: internal/tmux/tmux.go
package tmux

import (
	"fmt"
	"os/exec"
	"strings"
	"time"
)

func HasTmux() bool {
	_, err := exec.LookPath("tmux")
	return err == nil
}

func SessionExists(session string) bool {
	cmd := exec.Command("tmux", "has-session", "-t", session)
	return cmd.Run() == nil
}

func StartDetachedSession(session, command, cwd string) error {
	shellCmd := command
	if strings.TrimSpace(cwd) != "" {
		shellCmd = fmt.Sprintf("cd %s && %s", shellEscape(cwd), command)
	}
	cmd := exec.Command("tmux", "new-session", "-d", "-s", session, shellCmd)
	if out, err := cmd.CombinedOutput(); err != nil {
		msg := strings.TrimSpace(string(out))
		if msg == "" {
			msg = err.Error()
		}
		return fmt.Errorf("failed to start tmux session: %s", msg)
	}
	return nil
}

func shellEscape(s string) string {
	if s == "" {
		return "''"
	}
	return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
}

func SendCtrlC(session string) bool {
	cmd := exec.Command("tmux", "send-keys", "-t", session, "C-c")
	return cmd.Run() == nil
}

func StopSession(session string, grace time.Duration) {
	_ = SendCtrlC(session)
	deadline := time.Now().Add(grace)
	for time.Now().Before(deadline) {
		if !SessionExists(session) {
			return
		}
		time.Sleep(200 * time.Millisecond)
	}
	_ = exec.Command("tmux", "kill-session", "-t", session).Run()
}
